### PR TITLE
Implement Guild Stage Instance

### PR DESCRIPTION
**Pull request by SQKo**
- Add Stage Instance Repository
- Add Stage Instance Part
- Add Stage Instance into Guild
- Add Stage Instances event handlers

Requires: https://github.com/discord-php/DiscordPHP-Http/pull/8

- [ ] Add documentation

### CODE REVIEW

**Comment by SQKo** [comment]
Reopening this because I got it working now

File: src/Discord/Parts/Channel/StageInstance.php [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+use Discord\Parts\Guild\Guild;
+use Discord\Parts\Part;
+
+/**
+ * A Stage Instance holds information about a live stage. on a Discord guild.
+ *
+ * @property string  $id                    The unique identifier of the Stage Instance.
+ * @property string  $guild_id              The unique identifier of the guild that the stage instance associated to.
+ * @property Guild   $guild                 The guild that the stage instance associated to.
```
Comment (davidcole1340) [inline src/Discord/Parts/Channel/StageInstance.php:22]: `Guild|null`? Not sure if the guild is nullable here but on `getGuildAttribute()` it has `?Guild`

File: src/Discord/Parts/Channel/StageInstance.php [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+            'privacy_level' => $this->privacy_level,
+        ];
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function getRepositoryAttributes(): array
+    {
+        var_dump($this->channel_id);
```
Comment (davidcole1340) [inline src/Discord/Parts/Channel/StageInstance.php:101]:
```suggestion
```

File: src/Discord/WebSockets/Events/StageInstanceCreate.php [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+     * @inheritdoc
+     */
+    public function handle(Deferred &$deferred, $data): void
+    {
+        /** @var StageInstance */
+        $stage_instance = $this->factory->create(StageInstance::class, $data, true);
+
+        if ($guild = $this->discord->guilds->get('id', $stage_instance->guild_id)) {
+            $guild->stage_instances->push($stage_instance);
+            $this->discord->guilds->push($guild);
```
Comment (davidcole1340) [inline src/Discord/WebSockets/Events/StageInstanceCreate.php:30]:
```suggestion
```
Don't need this line, was a mistake I made throughout the library earlier. PHP objects are reference counted, so `->get()` will return the object reference AFAIK

File: src/Discord/WebSockets/Events/StageInstanceDelete.php [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+     * @inheritdoc
+     */
+    public function handle(Deferred &$deferred, $data): void
+    {
+        $stage_instance = $this->factory->create(StageInstance::class, $data);
+
+        if ($guild = $stage_instance->guild) {
+            $guild->stage_instances->pull($stage_instance->id);
+
+            $this->discord->guilds->push($guild);
```
Comment (davidcole1340) [inline src/Discord/WebSockets/Events/StageInstanceDelete.php:30]:
```suggestion
```

As above

File: src/Discord/WebSockets/Events/StageInstanceUpdate.php [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+     * @inheritdoc
+     */
+    public function handle(Deferred &$deferred, $data): void
+    {
+        $stage_instance = $this->factory->create(StageInstance::class, $data, true);
+
+        if ($guild = $this->discord->guilds->get('id', $stage_instance->guild_id)) {
+            $old = $guild->stage_instances->get('id', $stage_instance->id);
+            $guild->stage_instances->push($stage_instance);
+            $this->discord->guilds->push($guild);
```
Comment (davidcole1340) [inline src/Discord/WebSockets/Events/StageInstanceUpdate.php:30]:
```suggestion
```
As above

**Review by davidcole1340 — CHANGES_REQUESTED** [review]

**Review by valzargaming — APPROVED** [review]

**Pull request merged**